app/src/main/java/org/mozilla/fenix/components/metrics/GleanMetricsService.kt

```diff
@@ -475,7 +480,7 @@ class GleanMetricsService(private val context: Context) : MetricsService {
         /*
          * We cannot stop until we're done starting.
          */
-        runBlocking { starter.join(); }
+        runBlocking { gleanInitializer.join(); gleanSetStartupMetrics.join() }
```

Comment:
@mdboom so, I wasn't entirely sure and had to test, but looks like this can cause a dead lock (very unlikely though). We're blocking the main thread until other work on the main thread (`Glean.initialize`) is completed. Tested this by delaying initialization. This code was there before though so we don't need to block this PR on it, but should schedule a fix?

Reply:
Good catch, @csadilek.  Indeed, I don't think `getStartupMetrics` needs to run on the main thread.  Only `Glean.initialize` does.  So maybe put it on some other thread?

Reply:
@mdboom to be more precise: It's very unlikely because stop would have to happen *before* initialization is launched for this to deadlock. It can not happen *while* initialization is running, because both stop and initialization run on the main thread. Still wondering if we really need to do this though (wait for it to be started)?

Reply:
Looks like `setStartupMetrics` does mutate quite a bit of state that is not thread-safe. So, we can't just move it to a different thread. I think the change is fine as is, as the deadlocking case (stopping before starting) was there before. I am just wondering if we really need to join on the `gleanInitializer` and `gleanSetStartupMetric` when stopping? Isn't calling ` Glean.setUploadEnabled(false)` good enough?

Reply:
Ah, I see.  `Glean.setUploadEnabled()` is completely independent of initialization.  If Glean isn't initialized, it's just a flag.  If it were to be called in parallel with initialization, it would just take effect immediately after initialization (i.e. still resulting in no data being sent).  So it's probably completely safe just to remove this `runBlocking` altogether.

Reply:
@mdboom ah ok. In this case let's remove the `runBlocking` here and simply cancel the jobs if they are still running when we call `stop`.

Reply:
@mdboom Thanks for reviewing and working with @csadilek ! Can you take a look at this new version of the PR that we just pushed? Thanks!